Make a how to for creating a Sorting (#2944)
SpikeInterface/spikeinterface: doc/how_to/load_your_data_into_sorting.rst

Load Your Own Data into a Sorting
=================================

> Why make a :code:`Sorting`?

Reviewer: As usual, your why sections are great!

Author: @JoeZiminski edited it so hats off for his assist.

SpikeInterface contains pre-build readers for the output of many common sorters.
However, what if you have sorting output that is not in a standard format (e.g.
old csv file)? If this is the case you can make your own Sorting object to load
your data into SpikeInterface. This means you can still easily apply various
downstream analyses to your results (e.g. building correlograms or for generating
a :code:`SortingAnalyzer``).

The Sorting object is a core object within SpikeInterface that acts as a convenient
way to interface with sorting results, no matter which sorter was used to generate
them. **At a fundamental level it is a series of spike times and a series of labels

> for each spike along with some associated metadata.** Below, we will show you have

Reviewer: ```suggestion
for each unit and a sampling frequency for transforming frames to time.** Below, we will show you have
```
I would not mention the metadata as this is optional and you say at is core.

> to take your existing data and load it as a SpikeInterface :code:`Sorting`` object.

Reviewer: ```suggestion
to take your existing data and load it as a SpikeInterface :code:`Sorting` object.
```
Double tick

Reading a standard spike sorting format into a :code:`Sorting`
-------------------------------------------------------------

For most spike sorting output formats the :code:`Sorting` is automatically generated. For example one could do

.. code-block:: python

    from spikeinterface.extractors import read_phy

    # For kilosort/phy output files we can use the read_phy
    # most formats will have a read_xx that can used.
    phy_sorting = read_phy('path/to/folder')

> And voila you now have your :code:`Sorting` object generated and can use it for further analysis. For all the

Reviewer: ```suggestion
And voilà you now have your :code:`Sorting` object generated and can use it for further analysis. For all the
```
Dude : P

Author: I can't add the accents on my Windows keyboard..... but I appreciate your accuracy.

Reviewer: Just kidding here : )

current formats see :ref:`compatible_formats`.

Loading your own data into a :code:`Sorting`
-------------------------------------------

This :code:`Sorting` contains important information about your spike trains including:

  * spike times: the peaks of the extracellular potentials expressed in samples/frames these can
    be converted to seconds under the hood using the sampling_frequency

>   * spike labels: the neuron id for each spike, can also be called cluster ids or unit ids

Reviewer: I never thought about them this way. I always thought that the units have a label and a spike train but I think I had the "spike dictionary" representation too prominent in my mind. This makes more sense in the context of the spike vector.

    Stored as the :code:`unit_ids` in SpikeInterface
  * sampling_frequency: the rate at which the recording equipment was run at. Note this is the
    frequency and not the period. This value allows for switching between samples/frames to seconds

There are 3 options for loading your own data into a sorting object

With lists of spike trains and spike labels
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In this case we need a list of spike times unit labels, sampling_frequency and optional unit_ids
if you want specific labels to be used (in this case we only create the :code:`Sorting` based on
the requested unit_ids).

.. code-block:: python

    import numpy as np
    from spikeinterface.core import NumpySorting

    # in this case we are making a monosegment sorting
    # we have four spikes that are spread among two neurons
    my_sorting = NumpySorting.from_times_labels(
                                                times_list=[

>                                                     np.array([1000,12000,15000,22000])

Reviewer: ```suggestion
                                                    np.array([1000,12000,15000,22000])   # Note this are frames not times
```

Author: There's a typo in your suggestion so I'll fix this one myself. But it is a good note!

                                                    ],
                                                labels_list=[

>                                                     np.array([0,1,0,1])

Reviewer: ```suggestion
                                                    np.array(["a","b","a","b"])
```
I think have non number ids is more illustrative. What do you think?

                                                    ],

>                                                 sampling_frequency=30_000.0

Reviewer 2: ```suggestion
my_sorting = NumpySorting.from_times_labels(
    times_list=[
        np.array([1000,12000,15000,22000])   # Note these are samples/frames not times in seconds
        ],
    labels_list=[
        np.array(["a","b","a","b"])
        ],
    sampling_frequency=30_000.0
)
```

Hello, can you update the whitespace so that the stuff inside the function has one tab space. Both are Black compliant, this suggestion is more beautiful!

Author: Done.

                                                )

With a unit dictionary
^^^^^^^^^^^^^^^^^^^^^^

We can also use a dictionary where each unit is a key and its spike times are values.
This is entered as either a list of dicts with each dict being a segment or as a single
dict for monosegment. We still need to separately specify the sampling_frequency

.. code-block:: python

    from spikeinterface.core import NumpySorting

    my_sorting = NumpySorting.from_unit_dict(
                                             units_dict_list={
                                                              '0': [1000,15000],
                                                              '1': [12000,22000],
                                                              },

>                                              sampling_frequency=30_000.0

Reviewer 2: also here

                                           )

With Neo SpikeTrains
^^^^^^^^^^^^^^^^^^^^

Finally since SpikeInterface is tightly integrated with the Neo project you can create

> a sorting from :code:`Neo.SpikeTrain` objects. See :doc:`Neo documentation<neo:index>`` for more information on

Reviewer 2: There's a double ` at the end of the the Neo link, which breaks it. Should be
```
:doc:`Neo documentation<neo:index>`
```

Author: Fixed.

Author: I also want to say that this was completely the IDE's fault :P

using :code:`Neo.SpikeTrain`'s.

.. code-block:: python

    from spikeinterface.core import NumpySorting

    # neo_spiketrain is a Neo spiketrain object
    my_sorting = NumpySorting.from_neo_spiketrain_list(neo_spiketrain,
                                                       sampling_frequency=30_000.0)

> 

Reviewer 2: ```suggestion
my_sorting = NumpySorting.from_neo_spiketrain_list(
    neo_spiketrain,
    sampling_frequency=30_000.0
)
```

also whitespace here

.. code-block:: python

    import numpy as np
    from spikeinterface.core import NumpySorting

    # in this case we are making three-segment sorting
    # we have four spikes that are spread among two neurons
    # in each segment
    my_sorting = NumpySorting.from_times_labels(
                                                times_list=[
                                                    np.array([1000,12000,15000,22000]),
                                                    np.array([30000,33000, 41000, 47000]),
                                                    np.array([50000,53000,64000,70000]),
                                                    ],
                                                labels_list=[
                                                    np.array([0,1,0,1]),
                                                    np.array([0,0,1,1]),
                                                    np.array([1,0,1,0]),
                                                ],
                                                sampling_frequency=30_000.0
                                                )

> 

Reviewer 2: and whitespace here
